File: billing-domain/src/main/java/com/example/billing/domain/payment/Payment.java
package com.example.billing.domain.payment;

import com.example.billing.domain.order.OrderId;
import com.example.billing.domain.shared.Money;

import java.time.Clock;
import java.time.Instant;
import java.util.Objects;

/**
 * Payment 애그리거트.
 *
 * <p>PG 호출의 결과를 보존. {@link #idempotencyKey} 가 unique 하므로 같은 키로 중복 결제 시도 시
 * DB unique constraint 가 보호 (DB-level idempotency). Redis NX 와 함께 두 단계 방어 (ADR-0006).</p>
 */
public class Payment {

    private final PaymentId id;
    private final OrderId orderId;
    private final Money amount;
    private final PaymentMethod method;
    private final String idempotencyKey;
    private PaymentStatus status;
    private String pgTransactionId;
    private String errorCode;
    private String errorMessage;
    private final Instant createdAt;
    private Instant updatedAt;
    private long version;

    private Payment(PaymentId id, OrderId orderId, Money amount, PaymentMethod method,
                    String idempotencyKey, PaymentStatus status, String pgTransactionId,
                    String errorCode, String errorMessage,
                    Instant createdAt, Instant updatedAt, long version) {
        this.id = id;
        this.orderId = orderId;
        this.amount = amount;
        this.method = method;
        this.idempotencyKey = idempotencyKey;
        this.status = status;
        this.pgTransactionId = pgTransactionId;
        this.errorCode = errorCode;
        this.errorMessage = errorMessage;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.version = version;
    }

    public static Payment initiate(OrderId orderId, Money amount, PaymentMethod method,
                                   String idempotencyKey, Clock clock) {
        Objects.requireNonNull(orderId, "orderId");
        Objects.requireNonNull(amount, "amount");
        if (!amount.isPositive()) throw new IllegalArgumentException("amount must be positive");
        Objects.requireNonNull(method, "method");
        Objects.requireNonNull(idempotencyKey, "idempotencyKey");
        if (idempotencyKey.isBlank()) throw new IllegalArgumentException("idempotencyKey must not be blank");
        Instant now = clock.instant();
        return new Payment(PaymentId.newId(), orderId, amount, method, idempotencyKey,
                PaymentStatus.PENDING, null, null, null, now, now, 0L);
    }

    /** 영속 계층에서 복원. */
    public static Payment restore(PaymentId id, OrderId orderId, Money amount, PaymentMethod method,
                                  String idempotencyKey, PaymentStatus status, String pgTransactionId,
                                  String errorCode, String errorMessage,
                                  Instant createdAt, Instant updatedAt, long version) {
        return new Payment(id, orderId, amount, method, idempotencyKey, status,
                pgTransactionId, errorCode, errorMessage, createdAt, updatedAt, version);
    }

    public PaymentEvents.PaymentApproved approve(String pgTransactionId, Clock clock) {
        if (status != PaymentStatus.PENDING) {
            throw new IllegalStateException("payment must be PENDING to approve, was " + status);
        }
        this.status = PaymentStatus.APPROVED;
        this.pgTransactionId = pgTransactionId;
        this.updatedAt = clock.instant();
        return new PaymentEvents.PaymentApproved(id, orderId, amount, pgTransactionId, updatedAt);
    }

    public PaymentEvents.PaymentRejected reject(String errorCode, String errorMessage, Clock clock) {
        if (status != PaymentStatus.PENDING) {
            throw new IllegalStateException("payment must be PENDING to reject, was " + status);
        }
        this.status = PaymentStatus.FAILED;
        this.errorCode = errorCode;
        this.errorMessage = errorMessage;
        this.updatedAt = clock.instant();
        return new PaymentEvents.PaymentRejected(id, orderId, amount, errorCode, errorMessage, updatedAt);
    }

    // Getters
    public PaymentId id() { return id; }
    public OrderId orderId() { return orderId; }
    public Money amount() { return amount; }
    public PaymentMethod method() { return method; }
    public String idempotencyKey() { return idempotencyKey; }
    public PaymentStatus status() { return status; }
    public String pgTransactionId() { return pgTransactionId; }
    public String errorCode() { return errorCode; }
    public String errorMessage() { return errorMessage; }
    public Instant createdAt() { return createdAt; }
    public Instant updatedAt() { return updatedAt; }
    public long version() { return version; }
}

File: billing-domain/src/main/kotlin/com/example/billing/domain/payment/Payment.kt
package com.example.billing.domain.payment

import com.example.billing.domain.order.OrderId
import com.example.billing.domain.shared.Money
import java.time.Clock
import java.time.Instant

/**
 * Payment 애그리거트.
 *
 * PG 호출의 결과를 보존. [idempotencyKey] 가 unique 하므로 같은 키로 중복 결제 시도 시
 * DB unique constraint 가 보호 (DB-level idempotency). Redis NX 와 함께 두 단계 방어
 * (ADR-0006).
 *
 * **상태 라이프사이클**: PENDING → (APPROVED | FAILED) — 종착. terminal 상태로 한 번 들어가면
 * 다시 못 바뀜 (`approve` / `reject` 가 PENDING 검사). 같은 결제 ID 에 대한 PG 재시도는 별도
 * Payment 가 아닌 application 측 retry budget 으로 처리.
 *
 * record-style accessor (`id()` / `status()` / `pgTransactionId()` 등) 는 `@get:JvmName`
 * 으로 Java/Kotlin 양쪽 호출자 호환 유지.
 */
class Payment private constructor(
    @get:JvmName("id") val id: PaymentId,
    @get:JvmName("orderId") val orderId: OrderId,
    @get:JvmName("amount") val amount: Money,
    @get:JvmName("method") val method: PaymentMethod,
    @get:JvmName("idempotencyKey") val idempotencyKey: String,
    status: PaymentStatus,
    pgTransactionId: String?,
    errorCode: String?,
    errorMessage: String?,
    @get:JvmName("createdAt") val createdAt: Instant,
    updatedAt: Instant,
    @get:JvmName("version") val version: Long,
) {

    @get:JvmName("status")
    var status: PaymentStatus = status
        private set

    @get:JvmName("pgTransactionId")
    var pgTransactionId: String? = pgTransactionId
        private set

    @get:JvmName("errorCode")
    var errorCode: String? = errorCode
        private set

    @get:JvmName("errorMessage")
    var errorMessage: String? = errorMessage
        private set

    @get:JvmName("updatedAt")
    var updatedAt: Instant = updatedAt
        private set

    fun approve(pgTransactionId: String, clock: Clock): PaymentEvents.PaymentApproved {
        check(status == PaymentStatus.PENDING) {
            "payment must be PENDING to approve, was $status"
        }
        this.status = PaymentStatus.APPROVED
        this.pgTransactionId = pgTransactionId
        this.updatedAt = clock.instant()
        return PaymentEvents.PaymentApproved(id, orderId, amount, pgTransactionId, updatedAt)
    }

    fun reject(errorCode: String, errorMessage: String, clock: Clock): PaymentEvents.PaymentRejected {
        check(status == PaymentStatus.PENDING) {
            "payment must be PENDING to reject, was $status"
        }
        this.status = PaymentStatus.FAILED
        this.errorCode = errorCode
        this.errorMessage = errorMessage
        this.updatedAt = clock.instant()
        return PaymentEvents.PaymentRejected(id, orderId, amount, errorCode, errorMessage, updatedAt)
    }

    companion object {
        @JvmStatic
        fun initiate(
            orderId: OrderId,
            amount: Money,
            method: PaymentMethod,
            idempotencyKey: String,
            clock: Clock,
        ): Payment {
            require(amount.isPositive) { "amount must be positive" }
            require(idempotencyKey.isNotBlank()) { "idempotencyKey must not be blank" }
            val now = clock.instant()
            return Payment(
                id = PaymentId.newId(),
                orderId = orderId,
                amount = amount,
                method = method,
                idempotencyKey = idempotencyKey,
                status = PaymentStatus.PENDING,
                pgTransactionId = null,
                errorCode = null,
                errorMessage = null,
                createdAt = now,
                updatedAt = now,
                version = 0L,
            )
        }

        /** 영속 계층에서 복원. */
        @JvmStatic
        fun restore(
            id: PaymentId,
            orderId: OrderId,
            amount: Money,
            method: PaymentMethod,
            idempotencyKey: String,
            status: PaymentStatus,
            pgTransactionId: String?,
            errorCode: String?,
            errorMessage: String?,
            createdAt: Instant,
            updatedAt: Instant,
            version: Long,
        ): Payment = Payment(
            id, orderId, amount, method, idempotencyKey, status,
            pgTransactionId, errorCode, errorMessage, createdAt, updatedAt, version,
        )
    }
}
